hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionProcessor.java

```diff
@@ -60,59 +64,113 @@ public TransactionProcessor(
+        // Build the initial frame for the transaction
+        final var initialFrame = frameBuilder.buildInitialFrameWith(
+                transaction,
+                updater,
+                context,
+                config,
+                parties.sender().getAddress(),
+                parties.toAddress(),
+                gasCharges.intrinsicGas());
 
-            // Build the initial frame for the transaction
-            final var initialFrame = frameBuilder.buildInitialFrameWith(
-                    transaction,
-                    updater,
-                    context,
-                    config,
-                    parties.sender().getAddress(),
-                    parties.toAddress(),
-                    gasCharges.intrinsicGas());
-            // Return the result of running the frame to completion
-            return frameRunner.runToCompletion(
+        // Compute the result of running the frame to completion
+        final HederaEvmTransactionResult result;
+        try {
+            result = frameRunner.runToCompletion(
                     transaction.gasLimit(), initialFrame, tracer, messageCall, contractCreation);
-        } catch (final HandleException failure) {
-            return HederaEvmTransactionResult.abortFor(failure.getStatus());
+        } catch (ResourceExhaustedException e) {
+            return resourceExhaustionFrom(transaction.gasLimit(), context.gasPrice(), e.getStatus());
         }
+        // Adjust the pending commit based on the result
+        gasCharging.maybeRefundGiven(
+                transaction.unusedGas(result.gasUsed()),
+                gasCharges.relayerAllowanceUsed(),
+                parties.sender(),
+                parties.relayer(),
+                context,
+                updater);
+        initialFrame.getSelfDestructs().forEach(updater::deleteAccount);
+
+        // Returns this result if we can commit it without resource exhaustion, otherwise returns a fees-only result
+        return safeCommit(result, transaction, updater, feesOnlyUpdater, context);
+    }
+
+    private HederaEvmTransactionResult safeCommit(
+            @NonNull final HederaEvmTransactionResult result,
+            @NonNull final HederaEvmTransaction transaction,
+            @NonNull final HederaWorldUpdater updater,
+            @NonNull final Supplier<HederaWorldUpdater> feesOnlyUpdater,
+            @NonNull final HederaEvmContext context) {
+        try {
+            updater.commit();
+        } catch (ResourceExhaustedException e) {
+            // TODO - increment sender nonce and re-charge gas using feesOnlyUpdater
+            return resourceExhaustionFrom(transaction.gasLimit(), context.gasPrice(), e.getStatus());
+        }
+        return result;
     }
 
     private record InvolvedParties(
             @NonNull HederaEvmAccount sender, @Nullable HederaEvmAccount relayer, @NonNull Address toAddress) {}
 
     private InvolvedParties computeInvolvedParties(
```

Comment:
Now _this_ method _does_ need a method comment please!  It seems kind of intricate.

Reply:
Added javadoc and changed name to `setup()` since it can have a side-effect on the `updater` argument.